File: ocaml/src/problems_test.ml
open OUnit2 ;;
open Yojson ;;

let answer n =
    let json = Yojson.Basic.from_file "../../answers.json" in
    let open Yojson.Basic.Util in
    json |> member (string_of_int n) |> to_string

let ne_msg expected actual =
    "Expected " ^ expected ^ " but have " ^ actual

let test_problem01 context =
    let expected = answer 1 in
    let actual = Problems.problem01 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem02 context =
    let expected = answer 2 in
    let actual = Problems.problem02 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem03 context =
    let expected = answer 3 in
    let actual = Problems.problem03 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem04 context =
    let expected = answer 4 in
    let actual = Problems.problem04 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem05 context =
    let expected = answer 5 in
    let actual = Problems.problem05 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem06 context =
    let expected = answer 6 in
    let actual = Problems.problem06 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem07 context =
    let expected = answer 7 in
    let actual = Problems.problem07 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem08 context =
    let expected = answer 8 in
    let actual = Problems.problem08 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem09 context =
    let expected = answer 9 in
    let actual = Problems.problem09 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem10 context =
    let expected = answer 10 in
    let actual = Problems.problem10 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem11 context =
    let expected = answer 11 in
    let actual = Problems.problem11 () in
    assert_equal expected actual ~msg:(ne_msg expected actual)

let test_problem12 context =
    let expected = answer 12 in
    let actual = Problems.problem12 () in
    begin
        Printf.printf "Test 12 passed: got %s" actual;
        assert_equal expected actual ~msg:(ne_msg expected actual)
    end

let problem_tests = "test the problems" >:::
[
  "test problem 1" >:: test_problem01 ;
  "test problem 2" >:: test_problem02 ;
  "test problem 3" >:: test_problem03 ;
  "test problem 4" >:: test_problem04 ;
  "test problem 5" >:: test_problem05 ;
  "test problem 6" >:: test_problem06 ;
  "test problem 7" >:: test_problem07 ;
  "test problem 8" >:: test_problem08 ;
  "test problem 9" >:: test_problem09 ;
  "test problem 10" >:: test_problem10 ;
  "test problem 11" >:: test_problem11 ;
  "test problem 12" >:: test_problem12 ;
]

let _ = run_test_tt_main problem_tests

File: go/src/problems/problems_test.go
package problems

import (
	"encoding/json"
	"io/ioutil"
	"os"
	"testing"
)

var answers map[int]string

func TestMain(m *testing.M) {
	data, err := ioutil.ReadFile("../../../answers.json")
	if err != nil {
		panic(err)
	}

	err = json.Unmarshal(data, &answers)
	if err != nil {
		panic(err)
	}

	code := m.Run()
	os.Exit(code)
}

func TestProblem01(t *testing.T) {
	answer := Problem01()
	if answer != answers[1] {
		t.Error("Found", answer, "expected", answers[1])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem02(t *testing.T) {
	answer := Problem02()
	if answer != answers[2] {
		t.Error("Found", answer, "expected", answers[2])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem03(t *testing.T) {
	answer := Problem03()
	if answer != answers[3] {
		t.Error("Found", answer, "expected", answers[3])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem04(t *testing.T) {
	answer := Problem04()
	if answer != answers[4] {
		t.Error("Found", answer, "expected", answers[4])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem05(t *testing.T) {
	answer := Problem05()
	if answer != answers[5] {
		t.Error("Found", answer, "expected", answers[5])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem06(t *testing.T) {
	answer := Problem06()
	if answer != answers[6] {
		t.Error("Found", answer, "expected", answers[6])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem07(t *testing.T) {
	answer := Problem07()
	if answer != answers[7] {
		t.Error("Found", answer, "expected", answers[7])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem08(t *testing.T) {
	answer := Problem08()
	if answer != answers[8] {
		t.Error("Found", answer, "expected", answers[8])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem09(t *testing.T) {
	answer := Problem09()
	if answer != answers[9] {
		t.Error("Found", answer, "expected", answers[9])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem10(t *testing.T) {
	answer := Problem10()
	if answer != answers[10] {
		t.Error("Found", answer, "expected", answers[10])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem11(t *testing.T) {
	answer := Problem11()
	if answer != answers[11] {
		t.Error("Found", answer, "expected", answers[11])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}

func TestProblem12(t *testing.T) {
	answer := Problem12()
	if answer != answers[12] {
		t.Error("Found", answer, "expected", answers[12])
	} else {
		t.Log("Found " + answer + " successfully")
	}
}
